### PR TITLE
add nodegroups ext_pillar

### PR DESCRIPTION
**Pull request by ahammond**
### What does this PR do?

Adds the nodegroups ext_pillar
### What issues does this PR fix or reference?

https://github.com/saltstack/salt/issues/33125
### Tests written?

Yes

### CODE REVIEW

**Comment by damon-atkins** [comment]
Their might be a file "the filename" needs to be listed in, so it shows up in the doco.  ????

File: salt/pillar/nodegroups.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        - state_active
+        - country_US
+        - type_saltmaster
+
+Configuring Nodegroups Pillar
+-----------------------------
+
+.. code-block:: yaml
+
+    ext_pillar:
```
Comment (rallytime) [inline salt/pillar/nodegroups.py]: Is there supposed to be a value in `ext_pillar`? Or should the `extenstion_modules` part be bumped in a bit? Just making sure this example is correct/is valid yaml.

File: salt/pillar/nodegroups.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+'''
+=================
+Nodegroups Pillar
+=================
+
+Introspection: to which nodegroups does my minion belong?
+Provides a pillar with the default name of `nodegroups`
+which contains a list of nodegroups which match for a given minion.
+
```
Comment (rallytime) [inline salt/pillar/nodegroups.py]: Mind adding a `.. versionadded:: Carbon` directive here for clarity?

File: salt/pillar/nodegroups.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+__author__ = 'Andrew Hammond <andrew.george.hammond@gmail.com>'
+__copyright__ = 'Copyright (c) 2016 AnchorFree Inc.'
+__license__ = 'Apache License, Version 2.0'
+__version__ = '0.0.1'
+
+
+def ext_pillar(minion_id, pillar, pillar_name=None, *args, **kwargs):
+    pillar_name = pillar_name or 'nodegroups'
```
Comment (rallytime) [inline salt/pillar/nodegroups.py]: Can you please add some documentation here for what this function does, and what each of the args/kwargs represent?

I know some of that is listed above in the examples, but I think the more the merrier for clarity! :)

**Comment by rallytime** [comment]
Thanks for submitting this change @ahammond!! I made a couple of comments about some small changes, mainly for docs. 

I think a review from @cachedout would be good as well here.

File: salt/pillar/nodegroups.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+# Import Salt libs
+from salt.minion import Matcher
+
+# Import 3rd-party libs
+import salt.ext.six as six
+
+
+__author__ = 'Andrew Hammond <andrew.george.hammond@gmail.com>'
+__copyright__ = 'Copyright (c) 2016 AnchorFree Inc.'
+__license__ = 'Apache License, Version 2.0'
```
Comment (cachedout) [inline salt/pillar/nodegroups.py]: We'd prefer that instead of these three lines, you just add your name to the AUTHORS file.

**Comment by cachedout** [comment]
This is one of my favorite PRs in a while. This is a clever solution to the problem. I really like it.

**Comment by ahammond** [comment]
@cachedout comment applied. The cleverness came from @thatch45, I just wrote the code.

**Comment by scthi** [comment]
@ahammond I'm sorry to ask, but I can't get this to work for other minions beside the one  running on the salt-master. 

My `nodegroups` are correctly configured, something like `salt -N mygroup test.ping` works well. 

But the nodgroups ext_pillar applies all minions the groups of the master, as if the check is *always* using the master information instead of the current minion ones.

**Comment by ahammond** [comment]
@scthi I think I've only ever used this on the master.

**Comment by scthi** [comment]
@ahammond Thanks for the quick reply.

I found a solution, which works for me. The only change I made is the matching logic, which I grabbed from `salt/pillar/file_tree.py`:

```
#!/usr/bin/env python
# -*- coding: utf-8 -*-

'''
=================
Nodegroups Pillar
=================

Introspection: to which nodegroups does my minion belong?
Provides a pillar with the default name of `nodegroups`
which contains a list of nodegroups which match for a given minion.

.. versionadded:: Carbon

Command Line
------------

.. code-block:: bash
    salt-call pillar.get nodegroups
    local:
        - class_infra
        - colo_sj
        - state_active
        - country_US
        - type_saltmaster

Configuring Nodegroups Pillar
-----------------------------

.. code-block:: yaml

    extension_modules: /srv/salt/ext
    ext_pillar:
      - nodegroups:
          pillar_name: 'nodegroups'

'''

# Import futures
from __future__ import absolute_import

# Import Salt libs
from salt.utils.minions import CkMinions

# Import 3rd-party libs
import salt.ext.six as six

__version__ = '0.0.1'

def ext_pillar(minion_id, pillar, pillar_name=None):
    '''
    A salt external pillar which provides the list of nodegroups of which the minion is a member.

    :param minion_id: used to compound matching nodegroups
    :param pillar: provided by salt, but not used by nodegroups ext_pillar
    :param pillar_name: optional name to use for the pillar, defaults to 'nodegroups'
    :return: a dictionary which is included by the salt master in the pillars returned to the minion
    '''

    pillar_name = pillar_name or 'nodegroups'
    all_nodegroups = __opts__['nodegroups']
    nodegroups_minion_is_in = []
    for nodegroup_name in six.iterkeys(all_nodegroups):
        ckminions = CkMinions(__opts__)
        match = ckminions.check_minions(
            all_nodegroups[nodegroup_name],
            'compound')

        if minion_id in match:
            nodegroups_minion_is_in.append(nodegroup_name)

    return {pillar_name: nodegroups_minion_is_in}

```

@rallytime @cachedout Please let me know if you're interested in a PR.

**Pull request merged**